pkg/controller/infinispan/encryption.go

```diff
@@ -0,0 +1,177 @@
+package infinispan
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	v1 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v1"
+	config "github.com/infinispan/infinispan-operator/pkg/infinispan/configuration"
+	kube "github.com/infinispan/infinispan-operator/pkg/kubernetes"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	p12 "software.sslmate.com/src/go-pkcs12"
+)
+
+const (
+	EncryptClientCertPrefix      = "trust.cert."
+	EncryptClientCAName          = "trust.ca"
+	EncryptKeystoreName          = "keystore.p12"
+	EncryptKeystorePath          = ServerRoot + "/conf/keystore"
+	EncryptMountPath             = "/etc/encrypt"
+	EncryptTruststoreName        = "truststore.p12"
+	EncryptTruststorePassword    = "password"
+	EncryptTruststorePasswordKey = "truststore-password"
+)
+
+var ctx = context.TODO()
+
+func ConfigureServerEncryption(i *v1.Infinispan, c *config.InfinispanConfiguration, client client.Client) error {
+	if i.IsEncryptionDisabled() {
+		return nil
+	}
+
+	configureNewKeystore := func(c *config.InfinispanConfiguration) {
+		c.Keystore.CrtPath = EncryptMountPath
+		c.Keystore.Path = EncryptKeystorePath
+		c.Keystore.Password = "password"
+		c.Keystore.Alias = "server"
+	}
+
+	tlsSecretName := i.GetEncryptionSecretName()
+	if tlsSecretName == "" {
+		if i.IsClientCertEnabled() {
+			// It's not possible for client cert to be configured if no tls secret is provided by the user or via a
+			// encryption service, as no certificates are available to be added to the server's truststore, resulting
+			// in no client's being able to authenticate with the server.
+			return fmt.Errorf("Field 'CertSecretName' must be provided for '%s' or '%s' to be configured", v1.ClientCertAuthenticate, v1.ClientCertNone)
+		}
+		return nil
+	}
+
+	tlsSecret := &corev1.Secret{}
+	err := client.Get(ctx, types.NamespacedName{Namespace: i.Namespace, Name: tlsSecretName}, tlsSecret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("Secret %s for endpoint encryption not found.", tlsSecretName)
+		}
+		return fmt.Errorf("Error in getting secret %s for endpoint encryption: %w", tlsSecretName, err)
+	}
+
+	if i.IsEncryptionCertFromService() {
+		if strings.Contains(i.Spec.Security.EndpointEncryption.CertServiceName, "openshift.io") {
+			configureNewKeystore(c)
+			if i.IsClientCertEnabled() {
+				caPem, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt")
```

Comment:
@ryanemerson please pay attention, that service CA certificate is not permanent and will be rotating on OCP
https://docs.openshift.com/container-platform/4.7/security/certificate_types_descriptions/service-ca-certificates.html

Reply:
:+1: In that case the operator pod will be restarted and the `service-ca.crt` file will be reloaded.

Reply:
Yes, true, but if we will recreate our Secret every time service CA updated this should OOB as expected

Reply:
Hmm true. So we need the `config_controller` to reconcile when this file is updated. Is it possible to simply watch the file for changes and then trigger a reconcile? Or do you know of a better way that allows us to take some action on cert rotation?

Reply:
We can add the same logic as ConfgiMap has, for example store Secret hash as env variable and update it inside StatefulSet Pod template
Also, we can add notes for removing Secret in case of certificate rotation, but this probably should not work on managed service platform

Reply:
Is this something that restarts the ispn pods without advice?

Reply:
@ryanemerson I've identified another issue related to getting data from `/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt` as this will be limiting with test locally running operator and
- should be handled correctly
- getting CA cert from the Secret (where it's originally storing)

Reply:
@ryanemerson did you see comment above about local run and direct `/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt` file read?

Reply:
Yes, but I forgot about it :upside_down_face: 

> * getting CA cert from the Secret (where it's originally storing)

Do you know the name/namespace of the secret?

Reply:
The Secret name is placed inside operator ServiceAccount in the operator installation namespace (SA need to be created with the `make run-local`)
```
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{},"labels":{"name":"infinispan-operator"},"name":"infinispan-operator","namespace":"namespace-for-testing"}}
  creationTimestamp: "2021-04-28T12:13:37Z"
  labels:
    name: infinispan-operator
  name: infinispan-operator
  namespace: namespace-for-testing
  resourceVersion: "145278"
  selfLink: /api/v1/namespaces/namespace-for-testing/serviceaccounts/infinispan-operator
  uid: 02b025d5-59ec-4bfd-b2bf-150d6033b2b4
secrets:
- name: infinispan-operator-token-lvljf
```
For OCP secrets list will be bigger than one and name should start with the `infinispan-operator-token-<random letters>`
```
apiVersion: v1
data:
  ca.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN5RENDQWJDZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRJeE1EUXlOakV3TlRrMU1sb1hEVE14TURReU5ERXdOVGsxTWxvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBT0ljClBGRmQ2TUVPUU84SHlQTXJraUovcDZqbXlFMmlsVHpPcjlzY05SdmdCZGVMbFJrS2FGWjFKNTRjMndSTDkvT1EKUEJ1RFk4SEFsVVdUU01YOUsydmJ3QTlmbmJXMStlMC9MRUVhQlltMkU2NXJ5Y21tTDEyNHBRcllwNnJvS3NkcwpZM29IalB3T3hPM2FNU0FuY1Z6VFVnSUJxVnppWWNKb0Y0Wm9MSDVKSnF1QWEwaE03ZXpHMWJ2K2JTaFdMTjBMCjBVMllpaGJBUk1abjhKdm9tM1Y5RjN5akgrSFVVV1hUdnpOUnhrSjdJWUJqam44bElFOXF6NHJXalpNSHRkelEKeUJYY0c0cFFUdDBKQzE2bjZNM2dPb3pQVmwwZUNzNXVNYmZ3N21PY3pWMHZUTEo1c2ZMK21ibzdNYkRxRmViNAp0VkJmWnpwZ0JqOHIzWGx4eE0wQ0F3RUFBYU1qTUNFd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0RRWUpLb1pJaHZjTkFRRUxCUUFEZ2dFQkFGbkhhUFV4NU5EOWwycUxNM2o2bDdtYnVGQkIKdUs5MGlaa3VENS9aamd3bmF3MlhSWW9xOFpKMDRTZitJbWl1NVBkbURHbTlDNG5EY012WU9SWTJhcStMWnltQQoxYWVNdVpzL09aOFdEeGFBRnJXU2JEdmovSXE3ZEdDdGFqaUNkTzlXNzVWUVBrRkNDNHB5SmxDVmY0bEZ6U0kyCjVzbGFiMjRlQndaWTc5dXBtWTZnV3ZpVTRJK3Vkdm4xUGxZVXZIa1NVdVhMR2MwcVVFMkY5dlBrWmxTM0l5dnMKQmJuMmhPS0grZW03Mm9wa2FHb0FrVC8vMkRUUTNjc1FnRTBqL01SNmd1VC82ZWhuRjZvWUx3akYrd1ROQWNaQwo5M2hxQVFhbVQvVmVPeTNueHVpTHV3bjF2N3I1aXcya0w1ejhRdjVNRXp4WC9kTGlMT3czdHNOa2tBUT0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
  namespace: bmFtZXNwYWNlLWZvci10ZXN0aW5n
  token: ZXlKaGJHY2lPaUpTVXpJMU5pSXNJbXRwWkNJNklrMUhablJ4TURCa1lreGhRV1ZrVkdsZmJFTnhWMlUzTkZsTVZ6QlVSSEpYY1ZGNVRtZDZXalpoYTJzaWZRLmV5SnBjM01pT2lKcmRXSmxjbTVsZEdWekwzTmxjblpwWTJWaFkyTnZkVzUwSWl3aWEzVmlaWEp1WlhSbGN5NXBieTl6WlhKMmFXTmxZV05qYjNWdWRDOXVZVzFsYzNCaFkyVWlPaUp1WVcxbGMzQmhZMlV0Wm05eUxYUmxjM1JwYm1jaUxDSnJkV0psY201bGRHVnpMbWx2TDNObGNuWnBZMlZoWTJOdmRXNTBMM05sWTNKbGRDNXVZVzFsSWpvaWFXNW1hVzVwYzNCaGJpMXZjR1Z5WVhSdmNpMTBiMnRsYmkxc2RteHFaaUlzSW10MVltVnlibVYwWlhNdWFXOHZjMlZ5ZG1salpXRmpZMjkxYm5RdmMyVnlkbWxqWlMxaFkyTnZkVzUwTG01aGJXVWlPaUpwYm1acGJtbHpjR0Z1TFc5d1pYSmhkRzl5SWl3aWEzVmlaWEp1WlhSbGN5NXBieTl6WlhKMmFXTmxZV05qYjNWdWRDOXpaWEoyYVdObExXRmpZMjkxYm5RdWRXbGtJam9pTURKaU1ESTFaRFV0TlRsbFl5MDBZbVprTFdJeVltWXRNVFV3WkRZd016TmlNbUkwSWl3aWMzVmlJam9pYzNsemRHVnRPbk5sY25acFkyVmhZMk52ZFc1ME9tNWhiV1Z6Y0dGalpTMW1iM0l0ZEdWemRHbHVaenBwYm1acGJtbHpjR0Z1TFc5d1pYSmhkRzl5SW4wLlZrOGgwYXVZQVYxRFQwYkxzYy1KaWtPUmlHaTktVUZNbXZJRmdhWklXdmNxcHNiZVJCSjh5elJ2UEJEXzBaeXNxcHg0eGx2Vnc0R0d2eFJMT2RubWw3NmtfSTdKT2JKNE1tNWIxTGF4Q3BaZ2Q3VGgyLUV2TzdKOVBXaVc4eTlIOVdGYTR4RXYyd1dic3d1TUZvWTZVeG1iZUpaNHJOWXZLQVJTV3huRTIxRXpWQV9UbWtFb2ptcnZNem85UEY2YjV0NTU0MWZTRnpQRHhNSU4tVm81SW1mYk11ckFzNkFRSFgzaHVKUkhtZTM4b1AxVVpKeGtmVDVaNkU0Q1FRVWhoY1VBVXVUU2NGVjVod2NwWXlqSnAzWGZWNVR2T01Eci1CVks3clByT2hCZ3o1bDVacUpEUHBwWTJZU2ZHSVBybk9kcFV5cUkzQUpsX1IzZUp2VEJHdw==
kind: Secret
metadata:
  annotations:
    kubernetes.io/service-account.name: infinispan-operator
    kubernetes.io/service-account.uid: 02b025d5-59ec-4bfd-b2bf-150d6033b2b4
  creationTimestamp: "2021-04-28T12:13:37Z"
  name: infinispan-operator-token-lvljf
  namespace: namespace-for-testing
  resourceVersion: "145277"
  selfLink: /api/v1/namespaces/namespace-for-testing/secrets/infinispan-operator-token-lvljf
  uid: 9462f3b0-4bce-4370-b85e-43912d138f00
type: kubernetes.io/service-account-token
```
I think `ca.crt` is what you are looking for.

Reply:
Also this code never been tested as probably on k8s and locally running operator it should fails

Reply:
I have found a problem with the current approach when using `Type: Service`. The secret created via the `service.beta.openshift.io/serving-cert-secret-name:` annotation is of type `kubernetes.io/tls`, so it's not possible to add any fields other than `tls.key` and `tls.crt`. Therefore, we can't add the `trustore.p12` to this secret.

In order for truststores to work with `Type: Service`, we will need to use a separate secret to store the truststore. 

I think we have 2 options:

1. Create a secret using a consistent naming scheme, e.g. `<cr-name>-truststore`
2. Add an additional field to the spec, allowing users to specify the secret name. Maybe `endpointEncryption.ClientCertSecret: <some-name>`

I would like to reduce the amount of configuration required by the user, so ideally I would like to avoid adding another field to the spec WDYT @dmvolod @rigazilla @Crumby ?

Reply:
+1 for the option #1. Please also note, that's possible to mount files from the separate Secrets with different path using subPath option if needed.

Reply:
I can see a usage of opt 2 in case a user want to use the same truststore on multiple clusters, opt1 force him to duplicate the secret to match the name.

Reply:
That's a good point, however it's worth noting that the endpoint secret always has to be in the same namespace as the Infinispan CR. So this is only advantageous if multiple Infinispan CRs are deployed in the same namespace.

Reply:
It depends of the customer security requirements and how often customers will have multiple clusters inside one project, as namespace isolation one of the better feature for security control and other features less secure.

Reply:
Maybe we should go for option 2 after all .. I think we currently allow this for all secrets except the operator admin credentials, so it would be a more consistent experience and probably easier to document.

I would also update my proposal to use `endpointEncryption.ClientCertSecretName` to be consistent with the naming elsewhere.

Reply:
yeah, that use case is probably not so common, on the other hand users can just ignore the additional field if default naming is ok

Reply:
> If someone will try to share Secret between clusters conflict could happens and operator will fails with the SetControllerReference as this is just name.

@dmvolod Just raised a very good point. This PR makes it so that we set the owner ref on the secret to the Infinispan CR, this is necessary so that we can watch for changes to the secret in order to trigger a redeploy if it's updated. Unfortunately, this means that it's no longer possible to share the secret between clusters.

So I think we have two options:

1. Allow secrets to be shared between Infinispan CRs in the same namespace, but require manual cluster restarts if the secret is updated. Updates to keystores at runtime are not detected by the server, so it's not possible to simply update the secret.

2. Require secrets to be used per Infinispan CR only.

IMO 2 is the better option. As a keystore secret should probably be cluster specific anyway, I think the main downside is not being able to share the truststore secret. That being said, duplicating secret yaml definitions is trivial especially for automated deployements.

Reply:
+1 for option B. We need to add a validation to prevent error and add this limitations to the documentation. Also setting owner reference to the Secret will cause a garbage collection and customers should be aware about this in the documentation.

Reply:
+1 for option B. sharing things between CR could rise problems. good point @dmvolod !